File: src/main/java/jkmau5/alternativeenergy/util/BufferedTileEntity.scala
package jkmau5.alternativeenergy.util

import net.minecraft.world.World
import net.minecraft.tileentity.TileEntity
import net.minecraft.block.Block
import net.minecraftforge.common.ForgeDirection

/**
 * No description given
 *
 * @author jk-5
 */
object BufferedTileEntity {
  def makeBuffer(world: World, x: Int, y: Int, z: Int, loadUnloaded: Boolean): Array[BufferedTileEntity] = {
    val buffer = new Array[BufferedTileEntity](6)
    for(direction <- ForgeDirection.VALID_DIRECTIONS){
      buffer(direction.ordinal()) = new BufferedTileEntity(world, x + direction.offsetX, y + direction.offsetY, z + direction.offsetZ, loadUnloaded)
    }
    buffer
  }
}

final class BufferedTileEntity(private val world: World, private val x: Int, private val y: Int, private val z: Int, private val loadUnloaded: Boolean) {

  private var blockID = 0
  private var tileEntity: TileEntity = _
  private val tracker = new TimeTracker

  this.refresh()

  def refresh(){
    this.tileEntity = null
    this.blockID = 0
    if(!this.loadUnloaded && !this.world.blockExists(x, y, z)) return
    this.blockID = this.world.getBlockId(x, y, z)
    val block = Block.blocksList(this.blockID)
    if(block != null && block.hasTileEntity(world.getBlockMetadata(x, y, z))){
      this.tileEntity = world.getBlockTileEntity(x, y, z)
    }
  }

  def set(id: Int, tile: TileEntity){
    this.blockID = id
    this.tileEntity = tile
    this.tracker.markTime(world)
  }

  def getBlockID: Int = {
    if(this.tileEntity != null && !this.tileEntity.isInvalid) return this.blockID
    if(this.tracker.markIfTimePassed(this.world, 20)){
      this.refresh()
      if(this.tileEntity != null && !this.tileEntity.isInvalid) return this.blockID
    }
    0
  }

  def getTileEntity: TileEntity = {
    if(this.tileEntity != null && !this.tileEntity.isInvalid) return this.tileEntity
    if(this.tracker.markIfTimePassed(this.world, 20)){
      this.refresh()
      if(this.tileEntity != null && !this.tileEntity.isInvalid) return this.tileEntity
    }
    null
  }

  def exists: Boolean = {
    if(this.tileEntity != null && !this.tileEntity.isInvalid) true
    else this.world.blockExists(x, y, z)
  }
}

File: src/main/java/jkmau5/alternativeenergy/util/BufferedTileEntity.java
package jkmau5.alternativeenergy.util;

import net.minecraft.block.Block;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraftforge.common.ForgeDirection;

/**
 * No description given
 *
 * @author jk-5
 */
public class BufferedTileEntity {

    private int blockID = 0;
    private TileEntity tile;
    private final TimeTracker tracker = new TimeTracker();
    private final World world;
    final int x, y, z;
    private final boolean loadUnloaded;

    public BufferedTileEntity(World world, int x, int y, int z, boolean loadUnloaded) {
        this.world = world;
        this.x = x;
        this.y = y;
        this.z = z;
        this.loadUnloaded = loadUnloaded;
        this.refresh();
    }

    public final void refresh() {
        tile = null;
        blockID = 0;
        if (!loadUnloaded && !world.blockExists(x, y, z)) {
            return;
        }
        blockID = world.getBlockId(this.x, this.y, this.z);

        Block block = Block.blocksList[blockID];
        if (block != null && block.hasTileEntity(world.getBlockMetadata(this.x, this.y, this.z))) {
            tile = world.getBlockTileEntity(this.x, this.y, this.z);
        }
    }

    public void set(int blockID, TileEntity tile) {
        this.blockID = blockID;
        this.tile = tile;
        tracker.markTime(world);
    }

    public int getBlockID() {
        if (tile != null && !tile.isInvalid())
            return blockID;

        if (tracker.markTimeIfPassed(world, 20)) {
            refresh();

            if (tile != null && !tile.isInvalid())
                return blockID;
        }

        return 0;
    }

    public TileEntity getTile() {
        if (tile != null && !tile.isInvalid())
            return tile;

        if (tracker.markTimeIfPassed(world, 20)) {
            refresh();

            if (tile != null && !tile.isInvalid())
                return tile;
        }

        return null;
    }

    public boolean exists() {
        if(tile != null && !tile.isInvalid())
            return true;
        return world.blockExists(x, y, z);
    }

    public static BufferedTileEntity[] makeBuffer(World world, int x, int y, int z, boolean loadUnloaded) {
        BufferedTileEntity[] buffer = new BufferedTileEntity[6];
        for (int i = 0; i < 6; i++) {
            ForgeDirection d = ForgeDirection.getOrientation(i);
            buffer[i] = new BufferedTileEntity(world, x + d.offsetX, y + d.offsetY, z + d.offsetZ, loadUnloaded);
        }
        return buffer;
    }
}
